WIP EMP-389 prepare disk tests

raysiudak opened the pull request.

raysiudak committed:
Changes to alpine:egde because of bugs in f2fs-tools v1.10.x

fsck.f2fs SEGFAULTS, see EMP-375. When Alpine is updated to
v3.9 we should change back to alpine:latest

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

raysiudak committed:
Add file system checks after stub disk creation

We need to make sure the data we test with is correct.
Add some extra echo's so that there is better seperation
in the output.

Contributes to EMP-272, implements part of EMP-302

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

raysiudak committed:
Add prepare disk tests

- Test if the sha512 check returns with error
- Test to verify a boot partition resize, this test shares
  code with the grow rootfs test and therfore a helper function
  is created to avoid code duplicity.
- Test grow beyond disk size
- Test grow boot invalid start
- Test rootfs overlapping userdata
- Add random uint generator and made use of it in the test cases
- Removed the helper keyword from common functions
- Made the test_execute_disk_prepare_sha512_nok test more self
  explanatory.

Contributes to EMP-272, implements part of EMP-302

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

This branch requires a dockerimage update.

raysiudak committed:
Add disk integrity helper that can be used in the tests

    The helper function uses fdisk's validate functionality
    and fsck tools to validate the disk integrity.

    Contributes to EMP-272, implements part of EMP-302

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

oliv3r commented on build_for_ultimaker.sh:
```diff
@@ -1,5 +1,10 @@
 #!/bin/sh
-
```
excellent idea! but could you leave the whitespace (below the copyright text) for the vertically challenged ;

oliv3r commented on test/buildenv.sh:
```diff
@@ -1,5 +1,10 @@
 #!/bin/sh
-
```
ditto

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -51,9 +57,9 @@ ________________________________________________________________________________
-    mkfs.f2fs -q "${LOOP_STORAGE_DEVICE}p3"
+    mkfs.ext4 -q "${LOOP_STORAGE_DEVICE}p1" && fsck -y "${LOOP_STORAGE_DEVICE}p1"
+    mkfs.f2fs -q "${LOOP_STORAGE_DEVICE}p2" && fsck "${LOOP_STORAGE_DEVICE}p2"
+    mkfs.f2fs -q "${LOOP_STORAGE_DEVICE}p3" && fsck "${LOOP_STORAGE_DEVICE}p3"
```
i take it the f2fs tool doesn't have a -y argument? I'd also add the --force and the 'dry-run' option to ensure fsck is actually run

raysiudak replied:
Yes i will fix these in the next commit, because the function is introduced in that commit.

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -183,9 +189,12 @@ run_test()
+        echo
         echo "Result - OK"
     else
+        echo
```
de-duplicate these echo's, if you have the exact same code in both; maybe it's better to put it above. (Do we really need 2 blank lines before the result? (I'm ok with 2 blank lines)

raysiudak replied:
No we don't need it, but i think the output is nicer this way.

raysiudak replied:
We don't need this, but i think the output is nicer this way

oliv3r replied:
i'm okay with having the two new lines,

but don't put them both in the if/else, just put 2 echo's above the if; is what i meant:
```sh
echo
if "true"; then
   echo
   echo "result"
else
  echo
  echo "!result"
fi
```
would become
```sh
echo
echo
if "${1}"; then
   echo "Result - OK"
else
  echo "Result - error"
```
(or echo; echo; if you prefer it on a single line :p)

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -51,9 +57,9 @@ ________________________________________________________________________________
-    mkfs.f2fs -q "${LOOP_STORAGE_DEVICE}p3"
+    mkfs.ext4 -q "${LOOP_STORAGE_DEVICE}p1" && fsck -y "${LOOP_STORAGE_DEVICE}p1"
+    mkfs.f2fs -q "${LOOP_STORAGE_DEVICE}p2" && fsck "${LOOP_STORAGE_DEVICE}p2"
+    mkfs.f2fs -q "${LOOP_STORAGE_DEVICE}p3" && fsck "${LOOP_STORAGE_DEVICE}p3"
```
why not call the test integrity helper after formatting? isn't that what the helper is for? your duplicating its functionality here ;)

raysiudak replied:
Yes i will fix these in the next commit, because the function is introduced in that commit.

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -254,36 +263,124 @@ test_system_update_entrypoint()
 }
 
-test_execute_resize_partition_grow_rootfs_ok()
+helper_test_disk_integrity()
```
i'd almost argue this is not a helper, but just a regular function!

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -254,36 +263,124 @@ test_system_update_entrypoint()
 
     verification_partition_table_file="/tmp/verification_partition_table"
-    sfdisk -d "${LOOP_STORAGE_DEVICE}" > "${rootfs_dir}${verification_partition_table_file}"
+    sfdisk -d "${LOOP_STORAGE_DEVICE}" > "${rootfs_dir}${verification_partition_table_file}" || return 1
```
add the || return to the caller, or is that not possible due to the if [ "${test}" ]; because if that's the case, we really have to change the way we do that bit

raysiudak replied:
Ok it is a bit confusing. What i want in this function is that the result of executing prepare_disk is returned when it fails and if that the result is checked when it does not fail. Therefore i removed the '|| return' from the other test setup statements, because this is not right. The actual tests will fail if the test setup has errors.

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -254,36 +263,124 @@ test_system_update_entrypoint()
+    sed -i "s/p2.*type/p1 : start=     ${BOOT_START}, size=     ${new_boot_size}, type/" \
+        "${rootfs_dir}${PARTITION_TABLE_FILE}"
+
+    helper_execute_prepare_disk || return 0
```
seeing it here makes it even uglier imo

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -254,36 +263,124 @@ test_system_update_entrypoint()
+        "${rootfs_dir}${PARTITION_TABLE_FILE}"
+
+    helper_execute_prepare_disk || return 1
+    helper_test_disk_integrity || return 1
```
and now i want to strongly argue that they are not really 'helpers ' in that sense, but just regular functions we are calling

raysiudak replied:
Yeah they have grown to become a 'real' functions, i do agree.

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -254,36 +263,124 @@ test_system_update_entrypoint()
+{
+    new_boot_size="$((ROOTFS_START - BOOT_START + 1))"
+
+    sed -i "s/p2.*type/p1 : start=     ${BOOT_START}, size=     ${new_boot_size}, type/" \
```
document the regular expression so that we know what's happening in 2 years down the line
```sh
# Replace p2, the root partition with a larger p1, the boot partition
```
or something is enough and clear?

While this regex works quite well, i think a more complex one could have also helped by ONLY replacing the 'size' parameter value, rather then the entire line; but that's beyond my regex skills :p things like https://www.unix.com/shell-programming-and-scripting/172036-replace-specific-field-specific-line-sed-awk.html do have answer/solution, but they are not fun to read :p

I really think we need this because just glancing at it, I'm having trouble deciphering it. replace p2 with p1? huh?

raysiudak replied:
Yeah it was a typo

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -254,36 +263,124 @@ test_system_update_entrypoint()
 
+test_execute_disk_prepare_grow_boot_overlapping_rootfs_nok()
+{
+    new_boot_size="$((ROOTFS_START - BOOT_START + 1))"
```
While + 1 is okay-ish, why not do $((BOOT_SIZE + (${RANDOM} % ROOTFS_END) + 1)

to be fair, a proper test would then run this function RANDOM times to truly test it a few times.

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -254,36 +263,124 @@ test_system_update_entrypoint()
+{
+    new_rootfs_size="$((USERDATA_START - ROOTFS_START + 1))"
+
+    sed -i "s/p2.*type/p2 : start=     ${ROOTFS_START}, size=     ${new_rootfs_size}, type/" \
```
hmm, p2 again? was the previous one a typo?

raysiudak replied:
Yes

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -254,36 +263,124 @@ test_system_update_entrypoint()
+
+test_execute_disk_prepare_grow_boot_invalid_start_nok()
+{
+    new_boot_start="2047"
```
same here, 2048 is what we want, anything below 2048 is wrong, lets randomize it for broader test coverage.

If we feel crazy, we could even call this function 0 - 2047 times, but something more random is a bit safer

raysiudak replied:
Added a random value below 2048

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -254,36 +263,124 @@ test_system_update_entrypoint()
+
+test_execute_disk_prepare_grow_boot_ok()
+{
+    rootfs_size="$((USERDATA_START - ROOTFS_START))"
```
maybe use this pattern on the other functions as well?

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -254,36 +263,124 @@ test_system_update_entrypoint()
+test_execute_disk_prepare_grow_boot_ok()
+{
+    rootfs_size="$((USERDATA_START - ROOTFS_START))"
+    new_rootfs_size="$((rootfs_size / 2))"
```
while this is a good test, don't we want to maybe grow the rootfs as well? IN any case the function name doens't fully map to this. You only want to grow the boot. Granted to be able to do that somethings got to give.

raysiudak replied:
The test function says "test execution of disk_prepare by growing boot, expected result ok. The input is a partition table that should be correct, therefore the function modifies the boot "and" the rootfs partitions.

oliv3r replied:
i'm okay to keep as is, but the input partition could have the rootfs partition already reallocated to _only_ test the growing of boot of course. But as discussed, it's not worth the fixture overhead.

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -254,36 +263,124 @@ test_system_update_entrypoint()
+    partition_table_file="/tmp/partition_table"
+    sfdisk -d "${rootfs_dir}${STORAGE_DEVICE_IMG}" > "${rootfs_dir}${partition_table_file}"
+    chroot "${rootfs_dir}" sha512sum "${PARTITION_TABLE_FILE}" > "${rootfs_dir}${PARTITION_TABLE_FILE}.sha512"
+    echo "some extra data" >> "${rootfs_dir}${partition_table_file}"
```
note, that this test will fail because the 'extra data' will not validate, not because the shasum is invalid.

so i'd say, rewrite ($RANDOM) character(s) in the sha15sum?

raysiudak replied:
No im am making the partition file corrupt so that the md512sum does not validate anymore.

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -254,36 +263,124 @@ test_system_update_entrypoint()
+{
+    partition_table_file="/tmp/partition_table"
+    sfdisk -d "${rootfs_dir}${STORAGE_DEVICE_IMG}" > "${rootfs_dir}${partition_table_file}"
+    chroot "${rootfs_dir}" sha512sum "${PARTITION_TABLE_FILE}" > "${rootfs_dir}${PARTITION_TABLE_FILE}.sha512"
```
```sh
# Workaround to avoid having to deal with different workspaces, run sha512sum from within the ${rootfs_dir} workspace.
```

oliv3r replied:
could you add this comment above sha512sum still? to document why we are running it from within the chroot

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -341,7 +438,13 @@ run_test test_execute_resizef2fs
+run_test test_execute_disk_prepare_grow_rootfs_overlapping_userdata_nok
+run_test test_execute_disk_prepare_grow_boot_invalid_start_nok
+run_test test_execute_disk_prepare_grow_beyond_disk_end_nok
 run_test test_execute_resize_partition_grow_rootfs_ok
```
change on naming convention?

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -57,11 +66,17 @@ ________________________________________________________________________________
+        echo "Successfully created dummy storage device: '${LOOP_STORAGE_DEVICE}'."
+        return 0
+    fi
+
```
i'd prefer it if you'd swap these two around.

Test the integrity, and exit on failure (fail early). If everything is okay, then return normally (without error).

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -36,6 +36,11 @@ is_dev_setup_mounted=false
 
+random_uint32()
+{
+    echo "$(od -vAn -N4 -tu4 < /dev/urandom)"
```
while doing this briefly in a terminal as:
```sh
while true; do echo "$(od -vAn -N4 -tu4 < /dev/urandom)"; done
```
I would get many random numbers, but always very high. For example:
`
 3295899462
  849972328
  761581844
   55945594
`
So I'm questioning how 'random' this is. Will it ever return  < 10? I'm guessing that what happens is, we read 1 random byte after the other and then concatenate them? I thinking they basically are not uniformly distrbuted.

oliv3r replied:
[but busybox contains ](https://unix.stackexchange.com/a/140751) ...[shuf](https://linux.die.net/man/1/shuf)?!
```sh
shuf -i 0-"${MAX_RANGE}" -n 1
```

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -295,35 +300,45 @@ helper_execute_prepare_disk()
+    # Get a size between the current size and current + rootfs size
+    rootfs_size="$((USERDATA_START - ROOTFS_START))"
+    random_size="$(($(random_uint32) % rootfs_size))"
+    new_boot_size="$((BOOT_SIZE + random_size + 1))"
```
I  made a new friend today, his name is shuf! :D

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -295,35 +300,45 @@ helper_execute_prepare_disk()
-    new_rootfs_size="$((USERDATA_START - ROOTFS_START + 1))"
+    # Get a size between the current size and current + userdata size
+    userdata_size="$((STORAGE_DEVICE_SIZE - USERDATA_START))"
+    random_size="$(($(random_uint32) % userdata_size))"
```
just shuf it! :p

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -295,35 +300,45 @@ helper_execute_prepare_disk()
 test_execute_disk_prepare_grow_boot_invalid_start_nok()
 {
-    new_boot_start="2047"
+    new_boot_start="$(($(random_uint32) % $((BOOT_START -1))))"
```
missing space around the arithmetic operator - (unless you intended to say the number minus one?)

oliv3r commented on test/buildenv.sh:
```diff
@@ -11,7 +11,7 @@ ARM_EMU_BIN="${ARM_EMU_BIN:-}"
 
 FILESYSTEMS="ext4 f2fs overlay squashfs tmpfs"
-COMMANDS="apk mkfs.ext4 mkfs.f2fs mksquashfs sed sfdisk xz"
+COMMANDS="apk mkfs.ext4 mkfs.f2fs mksquashfs od sed sfdisk xz"
```
make shuf your new best friend too!

oliv3r commented on test/buildenv.sh:
```diff
@@ -96,10 +96,22 @@ check_command_installation()
+    fi
+    echo "exists"
+}
+
```
shuf don't care!

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -380,9 +380,9 @@ test_execute_disk_prepare_grow_boot_ok()
     # In every line in the partition table look for a string "p1<don't care>type" and replace it with
     # with new the new disk partition parameters defined above.
-    sed -i "s/p1.*type/p1 : start=     ${BOOT_START}, size=      ${new_boot_size}, type/" \
+    sed -i "s|${LOOP_STORAGE_DEVICE}p1.*type|${LOOP_STORAGE_DEVICE}p1 : start=        ${BOOT_START}, size=     ${new_boot_size}, type|" \
```
i assume the space count change before BOOT_START was intended?

oliv3r commented on build_for_ultimaker.sh:
```diff
@@ -1,5 +1,10 @@
+# Copyright (C) 2018 Olliver Schinagl <oliver@schinagl.nl>
+# Copyright (C) 2018 Raymond Siudak <raysiudak@gmail.com>
+#
+# SPDX-License-Identifier: AGPL-3.0+
```
its easier on the eyes to have 1 blank line between the end of the comment block and the first line of code

oliv3r commented on test/buildenv.sh:
```diff
@@ -1,12 +1,17 @@
+# Copyright (C) 2018 Raymond Siudak <raysiudak@gmail.com>
+# Copyright (C) 2018 Olliver Schinagl <oliver@schinagl.nl>
+#
+# SPDX-License-Identifier: AGPL-3.0+
```
ditto

oliv3r commented on scripts/prepare_disk.sh:
```diff
@@ -1,10 +1,9 @@
-#
-# SPDX-License-Identifier: AGPL-3.0+
-#
+
```
i can agree with you moving the license identifier below the copyright notices, that's good :+1: 
but keep it as one comment block:
```sh
#shebang
#
#stuff
#
# license :)
```

oliv3r commented on test/buildenv.sh:
```diff
@@ -1,5 +1,11 @@
 #!/bin/sh
 
```
ditto

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -1,5 +1,11 @@
 #!/bin/sh
 
+# Copyright (C) 2018 Ultimaker B.V.
```
ditto

oliv3r commented on build.sh:
```diff
@@ -1,4 +1,10 @@
+# Copyright (C) 2018 Olliver Schinagl <oliver@schinagl.nl>
+#
+# SPDX-License-Identifier: AGPL-3.0+
 
```
this block is the golden example :)

replicated it to stay concistent :)

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -29,6 +35,20 @@ result=0
+random_int()
+{
+    shuf -n 1 -i 0-"${1}"
+}
```
:100:

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -55,7 +75,13 @@ ________________________________________________________________________________
     mkfs.f2fs -q "${LOOP_STORAGE_DEVICE}p3"
 
+    if ! test_disk_integrity; then
+        echo "Something went wrong creating dummy storage device: '${LOOP_STORAGE_DEVICE}'."
```
Unknown error 50; something went wrong, Abort, Retry, Fail :p

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -278,12 +298,112 @@ test_execute_resize_partition_grow_rootfs_ok()
-	    -e   Stop consequtive tests on failure without cleanup
-	    -h   Print usage
+	    -e   Stop consecutive tests on failure without cleanup.
+	    -h   Print usage.
```
lol i just removed one period to stay consistent with 'ls --help' for example, where the options are not sentences.
So what will it be, are these instructive lines (like with ls --help, docker --help etc) or ... (i can't an example)?

you choose; i'll accept both solutions :)

raysiudak replied:
I choose this

raysiudak committed:
Add copyright notices

Contributes to EMP-272, implements part of EMP-302

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

This branch requires a dockerimage update.

raysiudak committed:
Fix grow boot and grow rootfs tests

- The grow partition tests now use a random size
- The 'sed' partition table string replace is no less errorprone by
  ignoring the witespace changes in the 'diff' compare

Contributes to EMP-272, implements part of EMP-302

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

raysiudak committed:
Add tests for correct resize on corrupted partitions

- Test disk_prepare with corrupted ext4 primary superblock
- Test disk prepare with corrupted f2fs primary superblock
- Test disk_prepare with corrupted f2fs superblocks

Contributes to EMP-272, EMP-302

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

raysiudak committed:
Error creating dummy loop device should be treated as unrecoverable.
When this happens, tests can not be run and a cleanup should be
triggered.
Additionally added a print of loop device partition in the setup.

Update dockerimage

Contributes to EMP-302

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

oliv3r committed:
scripts/prepare_disk: Do not query disk on each line

There is no need to get the disk-size for each line of processing.
Get it once at the start.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
scripts/prepare_disk: use long sfdisk arg

To improve readability and for consistency with sfdisk in the prepare
script, use long arguments. Note that this can break busybox sfdisk.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
scripts/prepare_disk: force formatting of partition

The mkfs.f2fs program sometimes fails to 'detect' the filesystem type
on a partition when the primary superblock is corrupted, but not
entirely and will refuse to run unless it is forced.

Since the script already made the decision to to format, so force it to
ensure the script can continue and not fail heron.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
scripts/prepare_disk: Add block device check

Our script should only work on block devices, if the expected partition
is not a block device, abort as we otherwise start formatting random
stuff.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
test/buildenv: Check for execute ability of arm emu

The build environment checks for an arm emulator with various tests,
but never tries to actually execute the emulator. By not doing this,
we can miss the fact that ARM_EMU_BIN exists as executable, but fails
to run due to missing shared libraries for example.

This patch adds a simple test to see whether the binary can actually
run.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
CI: also run test with the common parameters

With commit b4aa592dea4a ("Update CI configuration to trigger building in
temporary docker images") the common build paramaters (pkg_common then)
where accidentally dropped. This removed the tag 'docker' from the test
build job and would have precluded it from ever running.

With EMP-386 the CI environment was setup a new and the flag
'run-untagged' was rightfully removed. This caused tests to no longer
work however.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

raysiudak commented:
This pull request has been reworked and split up into multiple other pull requests
